core/unit_test/TestCudaCacheConfig.hpp

```diff
@@ -0,0 +1,114 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+#include <gtest/gtest.h>
+
+#include <Kokkos_Core.hpp>
+#include <iostream>
+
+// The following unit test is only valid for the CUDA backend.
+#if defined(KOKKOS_ENABLE_CUDA)
+namespace Test {
+TEST(TEST_CATEGORY, cuda_cache_config) {
+  using CachePreference = Kokkos::Impl::CachePreference;
+  CachePreference cache_config;
+  cudaFuncAttributes attributes;
+  cudaDeviceProp prop;
+  cudaGetDeviceProperties(&prop, 0);
```

Comment:
You don't technically need to use the CUDA runtime, you could directly set the fields that would care about.  That way you don't have to guard for Volta or Ampere below.

Reply:
The CUDA runtime call for properties is only to pass it to the function later on. I am not doing anything else with it. 
I purposefully set different field values based on architectures so that one set of values fail on the wrong architecture and hence had to protect them with guards. 
I can find a common set of filed values for both Volta and Ampere.

Reply:
It shouldn't be hard to find a value that works for Volta, Ampere, and Hopper (up to 228KB of shared memory, it's public info in the architecture overview at this point).

Reply:
I am saying pick one or more and do not bother with macro guards